src/server.rs

```diff
@@ -469,4 +473,112 @@ pub extern "C" fn rustls_server_session_get_sni_hostname(
+type ClientHelloUserData = *mut c_void;
+
+/// Called when the TLS ClientHello message was received and successfully decoded.
+type ClientHelloCallback = unsafe extern "C" fn(
+    userdata: ClientHelloUserData,
+    hello: *const rustls_client_hello,
+) -> rustls_result;
+
+struct ClientHelloResolver {
+    /// Implementation of rustls::ResolvesServerCert that passes values
+    /// from the supplied ClientHello to the callback function.
+    pub callback: ClientHelloCallback,
+    pub userdata: ClientHelloUserData,
+    pub previous_resolver: Option<Arc<dyn rustls::ResolvesServerCert>>,
+}
+
+impl ClientHelloResolver {
+    pub fn new(
+        callback: ClientHelloCallback,
+        userdata: ClientHelloUserData,
+        previous_resolver: Option<Arc<dyn rustls::ResolvesServerCert>>,
+    ) -> ClientHelloResolver {
+        ClientHelloResolver { callback, userdata, previous_resolver }
+    }
+}
+
+impl rustls::ResolvesServerCert for ClientHelloResolver {
+    fn resolve(&self, client_hello: ClientHello<'_>) -> Option<CertifiedKey> {
+        let sni_name = {
+            match client_hello.server_name() {
+                Some(c) => c.into(),
+                None => "",
+            }
+        };
+        let sigschemes = map_signature_schemes(client_hello.sigschemes());
+        let hello = rustls_client_hello {
+            sni_name: sni_name.as_ptr() as *const c_char,
+            sni_name_len: sni_name.len() as size_t,
+            signature_schemes: sigschemes.as_ptr(),
+            signature_schemes_len: sigschemes.len(),
+        };
+        let cb = self.callback;
+        let result: rustls_result = unsafe { cb(self.userdata, &hello) };
+        match result {
+            rustls_result::Ok => rustls_result::Ok,
+            r => panic!("client hello callback returned: {}", r)
+        };
+        match &self.previous_resolver {
+            Some(r) => r.resolve(client_hello),
+            None => None
+        }
+    }
+}
+
+unsafe impl marker::Sync for ClientHelloResolver {}
+unsafe impl marker::Send for ClientHelloResolver {}
+
+/// Register a callback to be invoked when a session created from this config
+/// is seeing a TLS ClientHello message. The given `userdata` will be passed
+/// to the callback when invoked.
+/// Specifying `replace`!= 0 will replace any existing `ResolvesServerCert` in
+/// the config. Otherwise, the existing `ResolvesServerCert` will be invoked
+/// after the callback has been returned successfully.
+/// Any error returned by the callback will abort the TLS handshake and give
+/// an error in the called rustls function (most likely
+/// `rustls_server_session_write_tls`).
+#[no_mangle]
+pub extern "C" fn rustls_server_config_builder_set_hello_callback(
+    builder: *mut rustls_server_config_builder,
+    callback: Option<extern "C" fn(
+        userdata: ClientHelloUserData,
+        hello: *const rustls_client_hello
+    ) -> rustls_result>,
```

Comment:
Let's use `Option<ClientHelloCallback>` (or `Option<rustls_client_hello_callback>`) here.

Reply:
I would, but then the generated `crustls.h` no longer contains `rustls_client_hello_callback` and realted things, but only an opaque

```
typedef struct Option_rustls_client_hello_callback Option_rustls_client_hello_callback;
```

Can one make `cbindgen` understand?